### ISSUE TITLE
Cross-listed Classes

### ISSUE THREAD

**Issue by willgearty** [learning-unlimited/ESP-Website#2180]
This feature was requested by a Stanford ESP teacher. They would like to be able to have a class listed in multiple categories if their class falls somewhere between them. I'm not even remotely sure how this would work (maybe pointers? similar to how you can use pointers to connect duplicate accounts?), so I'm just putting it here for the future.

**Comment by benjaminjkraft** [comment]
We could totally do this, via many-to-many fields!  I think it would be a pretty big change, and need to be made in a lot of places, since we often use the category in things like class codes.

**Comment by Guni102** [comment]
This is an interesting feature request.

If classes were cross-listed across multiple categories, would the intended approach be to move from a `ForeignKey` relationship to a `ManyToManyField` between classes and categories?

I’m curious how this would interact with places where categories are currently used directly (for example in class codes or category-based filtering). Would those continue to use a primary category, or would the system treat all associated categories equally?

Just trying to better understand the design implications here.

**Comment by willgearty** [comment]
We could still have a primary category (which is then used for the email code) and then have secondary categories?

Some other implementation questions that come to mind: 

- would we want to show the class separately under every category in the catalog?
- how would it count towards category capacity stats (e.g., the scheduling checks module) 

I'm sure there are other things to figure out too

**Comment by Guni102** [comment]
Thanks for the clarification,  keeping a primary category while allowing additional secondary categories makes sense.

One possible approach might be to keep the existing primary category as a ForeignKey (so current behavior like class codes, filtering, and email codes continues to work as it does now), while introducing a separate ManyToManyField for additional secondary categories. That would allow classes to appear in multiple categories while remaining backward-compatible with the current data model.

For catalog display, perhaps the class could appear under each associated category, but internally still reference the primary category for things like scheduling checks or category capacity statistics.

Alternatively, secondary categories might mainly serve discovery in the catalog, while reporting or admin views could still rely on the primary category to avoid ambiguity.

If this general direction seems reasonable, I’d be happy to help explore the implementation details further.

**Comment by willgearty** [comment]
Yeah, I like that approach! Means we don't need to convert any existing database fields, too.

**Comment by Guni102** [comment]
Hello! @willgearty 

I have opened a pull request #5022  addressing this issue. The PR includes the fix for the cross-listed class category behavior along with a brief explanation of the approach taken.

For verification, I tested the changes locally by running the ESP website using Docker, creating a test program, and configuring cross-listed class categories through the Django admin panel. Screenshots demonstrating the testing setup and results have also been added to the PR description and comments.

This issue was quite interesting to work on, and I learned a lot while understanding the catalog generation logic and debugging the mutation behavior in catalog_cached. Thank you for assigning this issue and for the opportunity to contribute.

Kindly take a look at the PR when you have time. Any feedback or suggestions would be greatly appreciated
